test_autofit/graphical/functionality/test_messages.py

```diff
@@ -98,4 +142,107 @@ def test_meanfield_gradients():
         norm = np.linalg.norm(grad[v] - njac1[v].sum((0, 1)))
         assert norm == pytest.approx(0, abs=1e-2)
         norm = np.linalg.norm(grad[v] - njac2[v].sum(0))
-        assert norm == pytest.approx(0, abs=1e-2)
\ No newline at end of file
+        assert norm == pytest.approx(0, abs=1e-2)
+
+
+def test_beta():
+    a = b = np.r_[2., 3.2, 1.5]
+    beta = graph.BetaMessage(a, b[::-1])
+    check_dist_norms(beta)
+
+    betas = [
+        graph.BetaMessage(a, b)
+        for a, b in (np.random.poisson(5, size=(10, 2)) + 1)
+    ]
+    [check_dist_norm(b) for b in betas]
+    check_log_normalisation(betas)
+
+
+def test_message_transforms():
+    a = b = np.r_[2., 3.2, 1.5]
+    mx, sigma = np.r_[-0.5, .1, 0.], np.r_[0.3, 0.5, 0.9]
+    test_cases = [
+        (graph.BetaMessage, (a, b)),
+        (graph.NormalMessage, (mx, sigma)),
+        (graph.UniformNormalMessage, (mx, sigma)),
+    ]
+    shift = (0.3, 5.1)
+    beta = graph.BetaMessage.shifted(0.3, 5.1)(a, b[::-1])
+
+
+def check_transforms(transform, x):
+    y, logd, logd_grad, jac = transform.transform_det_jac(x)
+
+    njac = numerical_jacobian(x, transform.transform)
+    nlogd_grad = numerical_jacobian(x, transform.log_det)
+
+    assert np.allclose(transform.inv_transform(y), x)
+    assert np.allclose(jac.to_dense(), njac)
+    assert np.isclose(logd.sum(), np.linalg.slogdet(njac)[1])
+    assert np.allclose(nlogd_grad.sum(0), logd_grad)
+
+
+def test_transforms():
+    tests = [
+        (transform.log_transform, np.r_[10, 2, 0.1]),
+        (transform.exp_transform, np.r_[10, -2, 0.1]),
+        (transform.logistic_transform, np.r_[0.22, 0.51, 0.1]),
+        (transform.phi_transform, np.r_[0.22, 0.51, 0.1]),
+        (transform.shifted_logistic(shift=11, scale=5.1), np.r_[11.1, 12, 16]),
+    ]
+    [check_transforms(*args) for args in tests]
+
+
+def test_multinomial_logit():
+    mult_logit = transform.multinomial_logit_transform
+
+    d = 3
+    p = np.random.dirichlet(np.ones(d+1))[:d]
+
+    x, logd, logd_grad, jac = mult_logit.transform_det_jac(p)
+
+    njac = numerical_jacobian(p, mult_logit.transform)
+    nlogd_grad = numerical_jacobian(p, mult_logit.log_det)
+
+    assert np.allclose(mult_logit.inv_transform(x), p)
+    assert np.allclose(jac.to_dense(), njac)
+    assert np.isclose(logd.sum(), np.linalg.slogdet(njac)[1])
+    assert np.allclose(nlogd_grad.sum(0), logd_grad)
+
+    n = 5
+
+    ps = np.random.dirichlet(np.ones(d+1), size=n)[:, :d]
```

Comment:
I'd probably make this a couple of tests?

If you have shared state you want to use in multiple tests then use fixtures

Reply:
using fixtures is annoying when debugging in Jupyter Notebooks, is there a way of making test set up easy to reuse in pytest and notebooks?

Reply:
Not that I know of. Best workflow might be to experiment in jupyter and then move to tests? As you know I'm more inclined to work entirely in tests

Reply:
Yeah I've currently written the tests so it's pretty painless to port them to Jupyter Notebook, I currently maintain a separate Jupyter notebook for the tests that don't port easily - in particular the regression examples, if we switch to using fixtures for these tests, I'd probably create another notebook that mirrors those tests, but I'd rather not as it would be easy for the notebook to diverge from the autofit tests...